### ISSUE TITLE
DE-EN WMT tutorial not giving expected output

### ISSUE THREAD

**Issue by sundeepteki** [awslabs/sockeye#638]
Hi,

I trained a DE-EN model exactly as per the instructions here - https://awslabs.github.io/sockeye/tutorials/wmt.html.

When I try to run the inference using the following command - 
echo "er ist so ein toller Kerl und ein Familienvater" | \
  python3 -m apply_bpe -c bpe.codes --vocabulary bpe.vocab.en \
                                   --vocabulary-threshold 50 | \
  python3 -m sockeye.translate -m wmt_model 2>/dev/null | \
  sed -r 's/@@( |$)//g'

The translation output I receive is: "it is is a a of of and and and and ,"
and not the expected output:         "he is a great guy and a family father ."

The final model details are as below -
"Training finished. Best checkpoint: 155. Best validation perplexity: 700.395322"

I wanted to check if my trained model parameters are consistent with those of the model that produced the output in the tutorial?

Thanks.

**Comment by tdomhan** [comment]
Could you paste the the last couple of lines from the training log and maybe the output of the `model/metrics` file?

Also if you have your exact training commands for double checking that would be helpful too.

**Comment by sundeepteki** [comment]
This is the output from log file - 
[2019-02-06:12:26:15:INFO:sockeye.training:collect_results] Decoder-187 finished: {'bleu-val': 0.003785395765721445, 'chrf-val': 0.11861884554235633, 'rouge1-val': 0.20614581305158247, 'rouge2-val': 0.019603614782491296, 'rougel-val': 0.1436156638114417, 'avg-sec-per-sent-val': 0.046578153133392336, 'decode-walltime-val': 23.289076566696167}
[2019-02-06:12:26:15:INFO:sockeye.training:fit] Training finished. Best checkpoint: 155. Best validation perplexity: 700.395322
[2019-02-06:12:26:15:INFO:sockeye.utils:__exit__] Releasing GPU 0.

Model/metrics file - 
187	avg-sec-per-sent-val=0.046578153133392336	bleu-val=0.003785395765721445	chrf-val=0.11861884554235633	decode-walltime-val=23.289076566696167	epoch=15	gradient-norm=1.2479667971962376	learning-rate=4.801999999999999e-05	perplexity-train=106.91554398656997	perplexity-val=713.993640141835	rouge1-val=0.20614581305158247 	rouge2-val=0.019603614782491296	rougel-val=0.1436156638114417	time-elapsed=37264.601153850555	used-gpu-memory=10493

And the training command - 
python3 -m sockeye.train -d train_data \
                        -vs newstest2016.tc.BPE.de \
                        -vt newstest2016.tc.BPE.en \
                        --encoder rnn \
                        --decoder rnn \
                        --num-embed 256 \
                        --rnn-num-hidden 512 \
                        --rnn-attention-type dot \
                        --max-seq-len 60 \
                        --decode-and-evaluate 500 \
                        --checkpoint-frequency 100 \
                        --device-ids 0 \
                        -o wmt_model

**Comment by sundeepteki** [comment]
Hi, any updates on this?

**Comment by mjpost** [comment]
That's a very high perplexity and similarly low validation BLEU score, so something has gone quite wrong. So something is quite wrong. What command did you use to prepare the training data? Maybe you had the language directions reversed?

**Comment by sundeepteki** [comment]
Indeed. I followed the tutorial to the t but I'll try to reproduce the error.

**Comment by mjpost** [comment]
That tutorial might also be outdated. You might also consider trying the tutorial under [Sockeye autopilot](https://github.com/awslabs/sockeye/tree/master/sockeye_contrib/autopilot).

**Comment by tdomhan** [comment]
This is quite strange indeed. Could you attach the entire log or paste the full beginning of the log (to sanity check vocab count etc). I will try rerunning the tutorial and see if I run into the same issues.

**Comment by tdomhan** [comment]
Did you limit the training data (the lines with `head -n 200000 ...`  in the tutorial?

**Comment by tdomhan** [comment]
I finally had some time to verify this and I am **not** able to reproduce the issue. Here are the steps I executed:

```
conda create -n sockeye_tutorial python=3.6

source activate sockeye_tutorial

pip install sockeye --no-deps -r requirements.gpu-cu92.txt

git clone https://github.com/rsennrich/subword-nmt.git
export PYTHONPATH=$(pwd)/subword-nmt:$PYTHONPATH

wget http://data.statmt.org/wmt17/translation-task/preprocessed/de-en/corpus.tc.de.gz
wget http://data.statmt.org/wmt17/translation-task/preprocessed/de-en/corpus.tc.en.gz
gunzip corpus.tc.de.gz
gunzip corpus.tc.en.gz
curl http://data.statmt.org/wmt17/translation-task/preprocessed/de-en/dev.tgz | tar xvzf -


python -m learn_joint_bpe_and_vocab --input corpus.tc.de corpus.tc.en \
                                    -s 30000 \
                                    -o bpe.codes \
                                    --write-vocabulary bpe.vocab.de bpe.vocab.en


python -m apply_bpe -c bpe.codes --vocabulary bpe.vocab.de --vocabulary-threshold 50 < corpus.tc.de > corpus.tc.BPE.de
python -m apply_bpe -c bpe.codes --vocabulary bpe.vocab.en --vocabulary-threshold 50 < corpus.tc.en > corpus.tc.BPE.en

python -m apply_bpe -c bpe.codes --vocabulary bpe.vocab.de --vocabulary-threshold 50 < newstest2016.tc.de > newstest2016.tc.BPE.de
python -m apply_bpe -c bpe.codes --vocabulary bpe.vocab.en --vocabulary-threshold 50 < newstest2016.tc.en > newstest2016.tc.BPE.en

python -m sockeye.prepare_data \
                        -s corpus.tc.BPE.de \
                        -t corpus.tc.BPE.en \
                        -o train_data

python -m sockeye.train -d train_data \
                        -vs newstest2016.tc.BPE.de \
                        -vt newstest2016.tc.BPE.en \
                        --encoder rnn \
                        --decoder rnn \
                        --num-embed 256 \
                        --rnn-num-hidden 512 \
                        --rnn-attention-type dot \
                        --max-seq-len 60 \
                        --decode-and-evaluate 500 \
                        -o wmt_model

# Training with a slightly smaller batch size:
 python -m sockeye.train -d train_data \
                        -vs newstest2016.tc.BPE.de \
                        -vt newstest2016.tc.BPE.en \
                        --encoder rnn \
                        --decoder rnn \
                        --num-embed 256 \
                        --rnn-num-hidden 512 \
                        --rnn-attention-type dot \
                        --max-seq-len 60 \
                        --decode-and-evaluate 500 \
                        -o wmt_model_smallbach --batch-size 2048

#Translate:

echo "er ist so ein toller Kerl und ein Familienvater ." | \
  python -m apply_bpe -c bpe.codes --vocabulary bpe.vocab.en \
                                   --vocabulary-threshold 50 | \
  python -m sockeye.translate -m wmt_model_smallbach 2>/dev/null | \
  sed -r 's/@@( |$)//g'

```
(note that I used a smaller batch size)

This produces:
```
> he is a a guy guy and a family father .
```

Final perplexities:
```
[2019-03-31:14:29:27:INFO:root:save_params_to_file] Saved params to "wmt_model_smallbach/params.00338"
[2019-03-31:14:29:27:INFO:sockeye.training:fit] Checkpoint [338]        Updates=1352000 Epoch=17 Samples=99552407 Time-cost=4751.748 Updates/sec=0.842
[2019-03-31:14:29:27:INFO:sockeye.training:fit] Checkpoint [338]        Train-perplexity=6.886707
[2019-03-31:14:29:46:INFO:sockeye.training:fit] Checkpoint [338]        Validation-perplexity=5.256161
[2019-03-31:14:29:53:INFO:sockeye.utils:log_gpu_memory_usage] GPU 0: 6192/11441 MB (54.12%)
[2019-03-31:14:29:53:INFO:sockeye.training:collect_results] Decoder-337 finished: {'bleu-val': 0.31160440885243934, 'chrf-val': 0.5338281700155819, 'rouge1-val': 0.5847193871188325, 'rouge2-val': 0.34614205046951774, 'rougel-val': 0.5197600801634908, 'avg-sec-per-sent-val': 0.14392298889160157, 'decode-walltime-val': 71.96149444580078}
[2019-03-31:14:29:53:INFO:sockeye.training:start_decoder] Starting process: Decoder-338
[2019-03-31:14:29:58:INFO:sockeye.training:fit] Validation-perplexity has not improved for 32 checkpoints, best so far: 5.249562
[2019-03-31:14:29:58:INFO:sockeye.lr_scheduler:new_evaluation_result] 8 checkpoints since improvement or rate scaling, lowering learning rate: 1.94e-06 -> 1.36e-06
[2019-03-31:14:30:04:INFO:sockeye.training:fit] Maximum number of not improved checkpoints (32) reached: 32
[2019-03-31:14:30:04:WARNING:sockeye.training:wait_to_finish] Waiting for process Decoder-338 to finish.
[2019-03-31:14:30:35:WARNING:sockeye.training:wait_to_finish] Had to wait 30 seconds for the Checkpoint Decoder-338 to finish. Consider increasing the checkpoint frequency (updates between checkpoints, see --checkpoint-frequency) or reducing the size of the validation samples that are decoded (see --decode-and-evaluate).
[2019-03-31:14:30:35:INFO:sockeye.training:collect_results] Decoder-338 finished: {'bleu-val': 0.31111721046826374, 'chrf-val': 0.5350197387551763, 'rouge1-val': 0.5875994688372096, 'rouge2-val': 0.3471943959830435, 'rougel-val': 0.5192682805429852, 'avg-sec-per-sent-val': 0.061370361328125, 'decode-walltime-val': 30.6851806640625}
[2019-03-31:14:30:35:INFO:sockeye.training:fit] Training finished. Best checkpoint: 306. Best validation perplexity: 5.249562
[2019-03-31:14:30:35:INFO:sockeye.utils:__exit__] Releasing GPU 0.
```

The discrepancy between the perplexities makes me wonder if maybe some files were swapped accidentally. Anyway, let me know if you have any more details for reproducing. Otherwise I will close the issue for now.

**Comment by tdomhan** [comment]
closing for now. Feel free to reopen.
